distributed/protocol/pickle.py

```diff
@@ -23,36 +24,39 @@ def _always_use_pickle_for(x):
         return False
 
 
-def dumps(x):
+def dumps(x, *, buffer_callback=None):
     """ Manage between cloudpickle and pickle
 
     1.  Try pickle
     2.  If it is short then check if it contains __main__
     3.  If it is long, then first check type, then check __main__
     """
+    dump_kwargs = {"protocol": HIGHEST_PROTOCOL}
+    if HIGHEST_PROTOCOL >= 5:
+        dump_kwargs["buffer_callback"] = buffer_callback
```

Comment:
Thinking a bit more, I think this current structure isn't correct. Passing in a buffer callback to this function can result in the buffer callback being called an incorrect number of times. For example, say we're pickling a tuple:

```
(object_that_supports_pickle_5, object_that_requires_cloudpickle)
```

With the current logic of trying pickle, then maybe trying cloudpickle, the following happens

- The buffer callback is used on the first call to `pickle.dumps` when it hits `object_that_supports_pickle_5`
- pickle errors when it hits `object_that_requires_cloudpickle`
- cloudpickle is then tried, resulting in the buffer callback being used on `object_that_supports_pickle_5` again
- The list of buffers contains erroneous items

To avoid that, I think you'll probably want to manage the list of buffers (and the callback) all internally in this file, and only return a list of frames.

Reply:
That's a fair concern. I'll take a look. Though am out this week. So it likely won't be until next week.

Reply:
Ok have pushed a fix to address this. Please let me know what you think 🙂